### PR TITLE
feature/login

### PR DESCRIPTION
**Pull request by GregVaz**
### Medium mode
- Sign in/ login
    - An user can register with username, email and password
    - An user can sign in 
    - An user can only see its books

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -59,3 +59,6 @@ typings/
 
 # next.js build output
 .next
+
+# sequelize migrations config
+config/config.json
```
Comment (jvargas98) [inline .gitignore]: missing empty line

File: app.mjs [hunk 2]
```diff
@@ -6,25 +6,70 @@ import { default as cookieParser } from 'cookie-parser';
...
 import { router as indexRouter } from './routes/index.mjs';
 import { router as booksRouter } from './routes/books.mjs';
+import { router as usersRouter } from './routes/users.mjs';
 import { default as rfs } from 'rotating-file-stream';
 import { default as DBG } from 'debug';
+import { default as passport } from 'passport';
+import { default as session } from 'express-session';
+import { Strategy as LocalStrategy } from 'passport-local'
+import dotenv from 'dotenv';
+dotenv.config();
```
Comment (jvargas98) [inline app.mjs:21]: empty line after import, please
```suggestion

dotenv.config();
```

File: controllers/books.mjs [hunk 3]
```diff
@@ -72,7 +76,8 @@ export async function destroyBook(req, res, next) {
     res.render('bookdestroy', {
       title: book ? book.title : "",
       id: book.id,
-      book: book
+      book: book,
```
Comment (jvargas98) [inline controllers/books.mjs]:
```suggestion
      book,
```
I think this would also work well

File: controllers/books.mjs [hunk 4]
```diff
@@ -86,4 +91,16 @@ export async function destroyBookConfirmation(req, res, next) {
...
+  try {
+    const book = await books.read(req.query.id);
+    if (req.user.email !== book.userId) {
+      res.redirect('/');
+    }
+    next();
+  } catch (err) {
+    res.redirect('/');
+  }
 }
```
Comment (jvargas98) [inline controllers/books.mjs]: missing empty line at the end of the file

File: controllers/index.mjs [hunk 5]
```diff
@@ -2,13 +2,32 @@ import { BooksStore as books } from '../models/books-store.mjs';
...
+  console.log(req.flash);
+  res.render('login', {
+    title: 'Login'
+  });
+}
+
+export function logout(req, res) {
+  req.logout();
+  res.redirect('/login');
 }
```
Comment (jvargas98) [inline controllers/index.mjs]: same here
```suggestion
}

```

File: controllers/users.mjs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+const debug = DBG('users:users-sequelize');
+const dberror = DBG('users:error-sequelize');
+
+export async function getUser(req, res, next) {
+  try {
+    let User = await users.read(req.query.email);
+    res.render('Userview', {
+      username: User ? User.username : '',
+      email: req.query.email,
+      User: User
```
Comment (jvargas98) [inline controllers/users.mjs]: I think this would also work well
```suggestion
      User,
```

File: controllers/users.mjs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  }
+}
+
+export async function updateUser(req, res, next) {
+  try {
+    const User = await users.read(req.query.email);
+    res.render('Useredit', {
+      username: User ? ("Edit " + User.username) : "Add a User",
+      email: User.email,
+      User: User
```
Comment (jvargas98) [inline controllers/users.mjs]: the same here
```suggestion
      User,
```

File: controllers/users.mjs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  } catch (err) { next(err); };
+}
+
+export async function destroyUser(req, res, next) {
+  try {
+    const User = await users.read(req.query.email);
+    res.render('Userdestroy', {
+      username: User ? User.username : "",
+      email: User.email,
+      User: User
```
Comment (jvargas98) [inline controllers/users.mjs]: same here

File: controllers/users.mjs [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+export async function destroyUserConfirmation(req, res, next) {
+  try {
+    await users.destroy(req.body.email);
+    res.redirect('/');
+  } catch (err) {
+    next(err);
+  }
+}
```
Comment (jvargas98) [inline controllers/users.mjs]: missing empty line

File: migrations/20210514220451-create-user.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+      },
+      username: {
+        type: Sequelize.STRING
+      }
+    });
+  },
+  down: async (queryInterface, Sequelize) => {
+    await queryInterface.dropTable('Users');
+  }
+};
```
Comment (jvargas98) [inline migrations/20210514220451-create-user.js:21]: the same here

File: models/books-sequelize.mjs [hunk 11]
```diff
@@ -20,12 +21,14 @@ async function connectDB() {
...
     abstract: Sequelize.DataTypes.TEXT,
-    cover: Sequelize.DataTypes.STRING
+    cover: Sequelize.DataTypes.STRING,
+    userId: Sequelize.DataTypes.STRING
   }, {
     sequelize,
     modelName: 'Books',
     timestamps: false
   });
+  // SQBook.belongsTo(SQUser);
```
Comment (jvargas98) [inline models/books-sequelize.mjs:31]: remove it if you no longer need it

File: models/users-sequelize.mjs [hunk 12]
```diff
@@ -0,0 +1,90 @@
+import { User, AbstractUsersStore } from './user.mjs';
+import { SQBook } from './books-sequelize.mjs';
+import Sequelize from 'sequelize';
+import {
+  connectDB as connectSequlz,
+  close as closeSequlz
+} from './index.mjs';
+import DBG from 'debug';
+const debug = DBG('users:users-sequelize');
```
Comment (jvargas98) [inline models/users-sequelize.mjs:9]: empty line between import and const, I'm not sure in this case

File: models/users-sequelize.mjs [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+      return new User(newUser.email, newUser.password, newUser.username); 
+    } 
+  }
+
+  async destroy(email) {
+    await connectDB();
+    await SQUser.destroy({ where: { email: email } });
+    debug(`DESTROY ${email}`);
+  }
+}
```
Comment (jvargas98) [inline models/users-sequelize.mjs]: empty line :)

**Review by jvargas98 — COMMENTED** [review]
Good job @GregVaz 🚀

File: routes/index.mjs [hunk 14]
```diff
@@ -1,6 +1,28 @@
 import { default as express } from 'express';
+import passport from 'passport';
 export const router = express.Router();
-import { getAllBooks } from '../controllers/index.mjs';
+import { getAllBooks, getRegister, getLogin, logout } from '../controllers/index.mjs';
 
-/* GET home page. */
-router.get('/', getAllBooks);
+router.get('/', (req, res, next) => {
+  if (req.isAuthenticated()) return next();
```
Comment (LoginALee) [inline routes/index.mjs]: This could be used as a middleware function in other file

**Review by LoginALee — APPROVED** [review]
Good job Goyo :robot:

File: .gitignore (hunk 1 above)
Comment (GregVaz) [reply .gitignore]: Yeah, it is missing :+1:

File: app.mjs (hunk 2 above)
Comment (GregVaz) [reply app.mjs:21]: Thanks, I'll check :+1:

File: controllers/books.mjs (hunk 3 above)
Comment (GregVaz) [reply controllers/books.mjs]: Yeah, it works :+1:

File: controllers/users.mjs (hunk 6 above)
Comment (GregVaz) [reply controllers/users.mjs]: Yes, it works, thanks :+1:

File: models/books-sequelize.mjs (hunk 11 above)
Comment (GregVaz) [reply models/books-sequelize.mjs:31]: Sure, just seems that it does not remove but in the next features I already remove it

File: models/users-sequelize.mjs (hunk 12 above)
Comment (GregVaz) [reply models/users-sequelize.mjs:9]: Thanks :D, probably also applies

File: routes/index.mjs (hunk 14 above)
Comment (GregVaz) [reply routes/index.mjs]: Yeah I need to structure better the project with middlewares

**Pull request merged**